tests/topotests/ospf-topo1/test_ospf_topo1.py

```diff
@@ -118,6 +116,78 @@ def teardown_module(mod):
     tgen.stop_topology()
 
 
+def test_wait_protocol_convergence():
+    "Wait for OSPFv2/OSPFv3 to converge"
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+
+    logger.info("waiting for protocols to converge")
+
+    def expect_ospfv2_neighbor_full(router, neighbor):
+        "Wait until OSPFv2 convergence."
+        logger.info("waiting OSPFv2 router '{}'".format(router))
+
+        def run_command_and_expect():
+            """
+            Function that runs command and expect the following outcomes:
+             * Full/DR
+             * Full/DROther
+             * Full/Backup
+            """
+            result = tgen.gears[router].vtysh_cmd('show ip ospf neighbor json',
+                                                  isjson=True)
+            if topotest.json_cmp(result, {"neighbors": {neighbor: [
+                    {"state": "Full/DR"}]}}) is None:
+                return None
+
+            if topotest.json_cmp(result, {"neighbors": {neighbor: [
+                    {"state": "Full/DROther"}]}}) is None:
+                return None
+
+            return topotest.json_cmp(result, {"neighbors": {neighbor: [
+                {"state": "Full/Backup"}]}})
+
+        _, result = topotest.run_and_expect(run_command_and_expect, None,
+                                            count=130, wait=1)
```

Comment:
is the one-second repeat timeout ... fine? it seems like minimal interval- is it possible these loops will run dozens of (or ... 100) times?

Reply:
it will call the function `run_command_and_expect` at most 130 times with a 1 second interval between each call. Each `run_command_and_expect` call will run `vtysh -c 'show ip ospf neighbor json`, then call the `json_cmp` three times at most (in case not `Full/DR` or `Full/DROther`).

the first time it returns `not None` it will cause the `assert` to trigger and quit the test with a failed status (it wont run the other `expect_ospfv2_neighbor_full` or `expect_ospfv3_neighbor_full` functions anymore).

Normally it converges within the first 11 seconds, worst case is it doesn't converge and it takes 130 seconds to return test failure (calling that function 130 times).

Does this explanation help? Do you think this is excessive (we have tons of tests doing this)?

Reply:
Sure, thanks - I just wanted to confirm that this was "customary" and ok...

Reply:
Yes... We don't have a better way to figure out when a protocol is "ready" other than polling. We could increase the interval but then we might spend more time waiting (because of the granularity change).

It should be possible to change this once we have northbound YANG notifications working: in that case we could open a connection with FRR and passively wait for the `state` to change to `Full`, but it is not there yet!